### PR TITLE
11.0 mig l10n it ddt

### PR DESCRIPTION
**Pull request by SimoRubi**
Superseding https://github.com/OCA/l10n-italy/pull/538

### CODE REVIEW

**Comment by SimoRubi** [comment]
Depends on https://github.com/OCA/stock-logistics-workflow/pull/447

**Comment by joezsweet** [comment]
any eta for this one?

**Comment by SimoRubi** [comment]
@joezsweet ETA on this depends on ETA on https://github.com/OCA/stock-logistics-workflow/pull/447, you could speed-up the process by reviewing that

**Comment by eLBati** [comment]
@joezsweet everybody can review.
See https://odoo-community.org/page/Contribute
and https://odoo-community.org/page/review

**Comment by joezsweet** [comment]
is this ready to merge?

**Comment by SimoRubi** [comment]
@joezsweet not yet.
I have just noticed that https://github.com/OCA/stock-logistics-workflow/pull/447 has been merged, then I'm now working on this

**Comment by SimoRubi** [comment]
Rebased on 11.0

**Comment by SimoRubi** [comment]
I think this is ready to be reviewed, @OCA/local-italy-maintainers could someone please check?
I know this module could use a little refactoring but I'd like to use this PR only for migrating, thanks

@joezsweet you can finally review this module :smile:

**Comment by joezsweet** [comment]
@simorubi where is possible to check functionality in runbot? Thx

**Comment by SimoRubi** [comment]
@joezsweet see https://drive.google.com/file/d/1JZhS2zwFK58MKNqUbZqFIxYfZMkd627g/view

**Comment by joezsweet** [comment]
@SimoRubi Thx, will have a look!

**Comment by joezsweet** [comment]
@SimoRubi have a look at: http://recordit.co/WZGRB41imu

Few things i noticed:
- I think shipping infos could be placed in above notebook, shown and set required if auto ddt creation is checked, imho will have a better ux.   
- Auto ddt creation could be moved to settings?
- What's the goal for ddt_type?
- Any chance to validate picking on "Set Done" in DDT if all products in picking have enough stock on hand for the selected warehouse? (aka reserved_qty = initial_demand)

What's the goal to have DDT and Package Preparation menu item?
schermata 2018-11-08 alle 17 59 52

Thx

**Comment by SimoRubi** [comment]
Thanks @joezsweet for your feedback!
This PR is strictly for migrating l10n_it_ddt from v10 to v11, so please just ensure that the behavior of the module here migrated to v11 is the same as it was in v10.

The changes you proposed here are all good but there is no difference from how it worked in v10.
You should open a new PR or a new issue in order to have them implemented.

**Comment by joezsweet** [comment]
@SimoRubi thx for the answer, will open a proposal issue. What about the issue in the video http://recordit.co/WZGRB41imu ?

**Comment by SimoRubi** [comment]
@joezsweet that is a nice catch! (I'm sorry but I didn't see the video earlier) 
I actually have managed to reproduce the issue in v10 and opened https://github.com/OCA/l10n-italy/issues/654 for what I think is the root cause.
I'll go to fix that in there and then I'll bring the fix in this PR.

Could you please make a review for this PR?

1. Go to the tab _Files changed_
2. Click on _Review changes_
3. Insert the link to the video (if you want add some explanation)
4. Select _Request changes_ and confirm

In this way it is easier for us to count positive/negative reviews (easier than reading all the comments in the thread).
Thanks

**Comment by eLBati** [comment]
@SimoRubi see https://github.com/OCA/l10n-italy/pull/670

**Comment by SimoRubi** [comment]
@eLBati cherry-picked in https://github.com/OCA/l10n-italy/pull/588/commits/803467499f08df4aab1fcd352cf011a0f0749d44

File: l10n_it_ddt/models/stock_picking_package_preparation.py [hunk 1]
```diff
@@ -287,19 +294,6 @@ def _prepare_invoice(self):
         """
         self.ensure_one()
         order = self._get_sale_order_ref()
-        if order:
```
Comment (eLBati) [inline l10n_it_ddt/models/stock_picking_package_preparation.py]: This would break integration with SO.
See https://github.com/OCA/l10n-italy/pull/542

**Comment by SimoRubi** [comment]
Thanks @glauco70 for your review!
I did what you asked for but I used `lot_ids` instead, it should be working anyway.

I also cleaned some unuseful modifications (most of the refactored strings) to their previous values in order to have a cleaner migration commit (after all the fix-ups).

**Comment by SimoRubi** [comment]
Cherry-picked https://github.com/OCA/l10n-italy/pull/656

**Comment by SimoRubi** [comment]
@joezsweet I have included changes from #656, could you please update your review (i.e. make a new one)?
Thanks

**Comment by glauco70** [comment]
@SimoRubi after the change in print DDT, there is a problem with "quantity_by_lot", the isn't this function in 'stock.picking.package.preparation.line', I try to add this from 10.0 but there is a problem with "self.move_id.quant_ids" : 'stock.move' object has no attribute 'quant_ids'

**Comment by SimoRubi** [comment]
@glauco70 I just tried to print the DDT report of a DDT having a product with serial number in runbot, see 
[DDT (11).pdf](https://github.com/OCA/l10n-italy/files/2638784/DDT.11.pdf).
Could you please describe the steps that raise the error?

**Comment by SimoRubi** [comment]
@joezsweet I see [here](https://github.com/OCA/l10n-italy/pull/588#pullrequestreview-179831060) that this is ok for you.
In order to give a positive review, you should do as described in https://github.com/OCA/l10n-italy/pull/588#issuecomment-437322910 and click on Approve in step 4.

Thanks!

**Comment by joezsweet** [comment]
@SimoRubi, few posts above ..

schermata 2018-12-03 alle 10 56 14

**Comment by SimoRubi** [comment]
@joezsweet I know, but if you look at the top of the PR:
image
on the right you can see yourself as a reviewer and the :x: means that in your last review you _Requested changes_: that is what bots and maintainers look at in order to count positive/negative reviews (they can't look at the whole conversation and see what you said inside your review).

**Comment by joezsweet** [comment]
@SimoRubi ok, done!

**Comment by SimoRubi** [comment]
Fixed up commits

**Comment by eLBati** [comment]
@SimoRubi please also squash the other commits

**Comment by SimoRubi** [comment]
@eLBati, squashed commits from same author

**Pull request merged**